plugins/plugin-java-debugger/che-plugin-java-debugger-ide/src/main/java/org/eclipse/che/plugin/jdb/ide/debug/JavaDebuggerFileHandler.java

```diff
@@ -49,128 +57,136 @@
+                            .then(new Function<ProjectConfigDto, JarFileNode>() {
+                                @Override
+                                public JarFileNode apply(ProjectConfigDto projectConfigDto) throws FunctionException {
+                                    return javaNodeFactory.newJarFileNode(jarEntry, null, projectConfigDto, nodeSettings);
+                                }
+                            })
+                            .then(new Operation<JarFileNode>() {
+                                @Override
+                                public void apply(final JarFileNode jarFileNode) throws OperationException {
+                                    AsyncCallback<VirtualFile> downloadSourceCallBack = new AsyncCallback<VirtualFile>() {
+                                        HandlerRegistration handler;
+
+                                        @Override
+                                        public void onSuccess(VirtualFile result) {
+                                            if (jarFileNode.isContentGenerated()) {
+                                                handler = eventBus.addHandler(TYPE, new FileContentUpdatedEventHandler() {
+                                                    @Override
+                                                    public void onFileSourceDownloaded(FileContentUpdatedEvent fileSourceDownloadedEvent) {
+                                                        handleActivateFile(jarFileNode, callback, location.getLineNumber());
+                                                        handler.removeHandler();
+                                                    }
+                                                });
+                                            } else {
+                                                handleActivateFile(jarFileNode, callback, location.getLineNumber());
+                                            }
+                                        }
+
+                                        @Override
+                                        public void onFailure(Throwable caught) {
+                                            callback.onFailure(caught);
+                                        }
+                                    };
+                                    handleActivateFile(jarFileNode, downloadSourceCallBack, location.getLineNumber());
+                                }
+                            })
+                            .catchError(new Operation<PromiseError>() {
+                                @Override
+                                public void apply(PromiseError arg) throws OperationException {
+                                    callback.onFailure(arg.getCause());
+                                }
+                            });
+    }
 
-        handleActivateFile(jarFileNode, callback);
-        eventBus.fireEvent(new FileEvent(jarFileNode, OPEN));
+    private String prepareFQN(String fqn) {
```

Comment:
As for me, I'd like to rename this method to something like `extractOuterClassFqn` or `getOuterClassFqn`